src/Stratis.Bitcoin.Features.PoA/FederationHistory.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+using NBitcoin.Crypto;
+using Stratis.Bitcoin.Features.PoA.Voting;
+
+namespace Stratis.Bitcoin.Features.PoA
+{
+    public interface IFederationHistory
+    {
+        /// <summary>Gets the federation member for a specified block.</summary>
+        /// <param name="chainedHeader">Identifies the block and timestamp.</param>
+        /// <returns>The federation member or <c>null</c> if the member could not be determined.</returns>
+        /// <exception cref="ConsensusErrorException">In case timestamp is invalid.</exception>
+        IFederationMember GetFederationMemberForBlock(ChainedHeader chainedHeader);
+
+        /// <summary>Gets the federation member for a specified block.</summary>
+        /// <param name="chainedHeader">Identifies the block and timestamp.</param>
+        /// <param name="federation">The federation members at the block height.</param>
+        /// <returns>The federation member or <c>null</c> if the member could not be determined.</returns>
+        IFederationMember GetFederationMemberForBlock(ChainedHeader chainedHeader, List<IFederationMember> federation);
+
+        /// <summary>Gets the federation for a specified block.</summary>
+        /// <param name="chainedHeader">Identifies the block and timestamp.</param>
+        /// <returns>The federation member or <c>null</c> if the member could not be determined.</returns>
+        List<IFederationMember> GetFederationForBlock(ChainedHeader chainedHeader);
+    }
+
+    public class FederationHistory : IFederationHistory
+    {
+        private readonly IFederationManager federationManager;
+        private readonly VotingManager votingManager;
+        private readonly Dictionary<uint256, IFederationMember> minersByBlockHash;
+
+        public FederationHistory(IFederationManager federationManager, VotingManager votingManager = null)
```

Comment:
We need to investigate why voting manager would be null as per new implementation between the daemons, it should always have an instance? Thoughts?

Reply:
Just keeping changes to a minimum but definitely worth considering.